Document release dates in changelog

jondricek opened the pull request:
<!-- Each pull request must be associated with an open issue unless it's a documentation fix. If a corresponding issue does not exist, please create one so we can discuss the change first. -->

<!-- Please answer the following. We close pull requests that don't. -->

> Which issue, if any, is this issue related to?

Closes #8259 

> Is there anything in the PR that needs further explanation?

I hope it is self explanatory! I grabbed the dates from https://www.npmjs.com/package/stylelint?activeTab=versions and feel like it would be useful to have them available

jeddy3 commented:
@jondricek Thank you for creating the PR. Dates in the changelog would be useful. I can't remember if there was ever a reason we didn't adopt them, especially as we tend to lean on [Keep a changelog](https://keepachangelog.com/en/1.1.0/) for changelog guidance and that has them.

We use [changesets](https://github.com/changesets/changesets) to generate our changelog entries. See https://github.com/stylelint/stylelint/pull/8202 for an example. Do you know if we can add the date automatically, or would it be a manual step before merging the changesets PR?

Mouvedia commented on CHANGELOG.md:
```diff
@@ -226,23 +226,23 @@
-## 16.0.0
+## 16.0.0 (2023-12-08)
 
 [Migrating to `16.0.0` guide](docs/migration-guide/to-16.md).
```
It should probably be

```md
## 16.0.0 (2023-12-08)

[Migrating to `16.0.0` guide](docs/migration-guide/to-16.md).

- Removed: Node.js less than 18.12.0 support ([#7020](https://github.com/stylelint/stylelint/pull/7020)) ([@ybiquitous](https://github.com/ybiquitous)).
- Removed: 76 rules deprecated in 15.0.0 ([#6979](https://github.com/stylelint/stylelint/pull/6979)) ([@mattxwang](https://github.com/mattxwang)).
- Changed: `stylelint.formatters` object to have `Promise` values ([#7184](https://github.com/stylelint/stylelint/pull/7184)) ([@ybiquitous](https://github.com/ybiquitous)).
- Changed: exit code for CLI flag error ([#7134](https://github.com/stylelint/stylelint/pull/7134)) ([@ybiquitous](https://github.com/ybiquitous)).
- Changed: default syntax behaviour to always use safe-parser with `fix` regardless of extension ([#7357](https://github.com/stylelint/stylelint/pull/7357)) ([@ybiquitous](https://github.com/ybiquitous)).
- Changed: TypeScript definitions for ESM ([#7309](https://github.com/stylelint/stylelint/pull/7309)) ([@ybiquitous](https://github.com/ybiquitous)).
- Changed: `stylelint.rules` object to have `Promise` values ([#7279](https://github.com/stylelint/stylelint/pull/7279)) ([@ybiquitous](https://github.com/ybiquitous)).
- Changed: `stylelint.utils.checkAgainstRule` to be an async function ([#7339](https://github.com/stylelint/stylelint/pull/7339)) ([@ybiquitous](https://github.com/ybiquitous)).
- Deprecated: CommonJS Node.js API ([#7353](https://github.com/stylelint/stylelint/pull/7353)) ([@ybiquitous](https://github.com/ybiquitous)).
- Deprecated: `output` property in a Node.js API returned object. Instead, `report`/`code` properties are recommended ([#7183](https://github.com/stylelint/stylelint/pull/7183)) ([@ybiquitous](https://github.com/ybiquitous)) & ([@haocheng6](https://github.com/haocheng6)).
- Added: `exports` field to `package.json` for Conditional Exports (ESM/CommonJS) ([#7307](https://github.com/stylelint/stylelint/pull/7307)) ([@ybiquitous](https://github.com/ybiquitous)).
- Added: support for ESM plugins ([#7339](https://github.com/stylelint/stylelint/pull/7339)) ([@ybiquitous](https://github.com/ybiquitous)).
- Added: support for ESM custom formatters ([#7343](https://github.com/stylelint/stylelint/pull/7343)) ([@ybiquitous](https://github.com/ybiquitous)).
- Added: support for a package name as arguments to `--custom-formatter` ([#7343](https://github.com/stylelint/stylelint/pull/7343)) ([@ybiquitous](https://github.com/ybiquitous)).
- Added: `severity` secondary option's function support ([#7202](https://github.com/stylelint/stylelint/pull/7202)) ([@kizu](https://github.com/kizu)).
- Added: support for a `Promise` formatter function ([#7184](https://github.com/stylelint/stylelint/pull/7184)) ([@ybiquitous](https://github.com/ybiquitous)).
- Added: support for ESM custom syntaxes ([#7351](https://github.com/stylelint/stylelint/pull/7351)) ([@ybiquitous](https://github.com/ybiquitous)).
- Added: `report`/`code` properties to a Node.js API returned object ([#7183](https://github.com/stylelint/stylelint/pull/7183)) ([@ybiquitous](https://github.com/ybiquitous)) & ([@haocheng6](https://github.com/haocheng6)).
- Fixed: no longer needed workaround for Cosmiconfig segfault ([#7329](https://github.com/stylelint/stylelint/pull/7329)) ([@ybiquitous](https://github.com/ybiquitous)).
- Fixed: CLI regression to avoid waiting for stdin without any input ([#7131](https://github.com/stylelint/stylelint/pull/7131)) ([@ybiquitous](https://github.com/ybiquitous)).
- Fixed: CLI to avoid different outputs on empty files and empty stdin ([#7131](https://github.com/stylelint/stylelint/pull/7131)) ([@ybiquitous](https://github.com/ybiquitous)).
- Refactored: `.js` extension to `.mjs` and `.cjs` ([#7307](https://github.com/stylelint/stylelint/pull/7307)) ([@ybiquitous](https://github.com/ybiquitous)).

## 16.0.0-0–16.0.0-3 (2023-11-08–2023-11-27)

- Changed: `stylelint.formatters` object to have `Promise` values ([#7184](https://github.com/stylelint/stylelint/pull/7184)) ([@ybiquitous](https://github.com/ybiquitous)).
- Changed: default syntax behaviour to always use safe-parser with `fix` regardless of extension ([#7357](https://github.com/stylelint/stylelint/pull/7357)) ([@ybiquitous](https://github.com/ybiquitous)).
- Changed: `stylelint.utils.checkAgainstRule` to be an async function ([#7339](https://github.com/stylelint/stylelint/pull/7339)) ([@ybiquitous](https://github.com/ybiquitous)).
- Deprecated: CommonJS Node.js API ([#7353](https://github.com/stylelint/stylelint/pull/7353)) ([@ybiquitous](https://github.com/ybiquitous)).
- Added: support for ESM plugins ([#7339](https://github.com/stylelint/stylelint/pull/7339)) ([@ybiquitous](https://github.com/ybiquitous)).
- Added: support for ESM custom formatters ([#7343](https://github.com/stylelint/stylelint/pull/7343)) ([@ybiquitous](https://github.com/ybiquitous)).
- Added: `exports` field to `package.json` for Conditional Exports (ESM/CommonJS) ([#7307](https://github.com/stylelint/stylelint/pull/7307)) ([@ybiquitous](https://github.com/ybiquitous)).
- Fixed: no longer needed workaround for Cosmiconfig segfault ([#7329](https://github.com/stylelint/stylelint/pull/7329)) ([@ybiquitous](https://github.com/ybiquitous)).
```

jondricek commented:
I've been nerd sniped! I just read through all the documentation on changesets. Probably should have done that before since I just wandered in from the cold and started messing with your changelog and you of course already had a process in place. According to this 5+ year old open issue (https://github.com/changesets/changesets/issues/109), it looks like there is appetite from some users to have this (there's even a 4 year old [open PR](https://github.com/changesets/changesets/pull/400)), but everything is sort of left up to the user to customize their own experience.

Lucky for you, you are already customizing your process by running the rewrite-changelog.mjs file (https://github.com/stylelint/stylelint/blob/main/package.json#L81C19-L81C36)!  I just updated my branch with

- made the dates 100% compliant with Keep a Changelog
- updated the [rewrite-changelog.mjs](https://github.com/stylelint/stylelint/pull/8260/files#diff-63608a9438433dcce04b09e6d352dd2729d4c07c682645c06b86d87f135c6824) file to add the date, but only to the latest version, keeping the rest of the file untouched
    - Tested this a couple times on my computer... but not sure how to best incorporate a test for this into the overall process in this repo. It feels a bit out of the normal scope of the tests for this repo. Its only a 3 line fix, so hopefully straightforward to read over and test yourselves as well

jondricek commented:
@Mouvedia I see your comment, but I'm not exactly sure what to do on my branch about it. I'm happy to add in the section for 

```
## 16.0.0-0–16.0.0-3 - 2023-11-08 – 2023-11-27
...
<content here>
...
```

for you, but it looks like that whole section is missing from the Changelog at the moment, right?

Also, other than manually doing that for historic entries, I'm not sure how to codify that into the script or anything.

jeddy3 commented:
Thank you for diving into the changesets docs! I'm glad we can automate adding the date.

The only other piece of the puzzle I'm aware of is how we generate the [GitHub release](https://github.com/stylelint/stylelint/releases) from the changelog using [our custom GitHub Action](https://github.com/stylelint/changelog-to-github-release-action/blob/main/src/changelogToGithubRelease.js). You're welcome to take a look. If it's not obvious if it requires tweaking, @ybiquitous may be able to advise, as he's most familiar with the mechanics of our release process.

As an aside, it's interesting to see the release date of our first version as 2015-07-08. I had forgotten that it took us six months to prepare our first release after creating [our first issue](https://github.com/stylelint/stylelint/issues/1) in December of the previous year, which we recently referenced as Stylelint turning 10 years old.

jondricek commented:
Ok. From what i'm seeing in this setup:

- The [Preparing GitHub Release](https://github.com/stylelint/stylelint/actions/workflows/preparing-github-release.yml) action [only happens for git tags](https://github.com/stylelint/stylelint/blob/main/.github/workflows/preparing-github-release.yml#L3-L6)
- The custom GitHub Action compares the git tag as the variable `version` to the Markdown header as `text.value` ([link to line](https://github.com/stylelint/changelog-to-github-release-action/blob/main/src/changelogToGithubRelease.js#L15))
    - This has worked up to this point, but my PR breaks it by adding the date
    - I think if that line's if statement was updated to be `(text?.type === 'text' && text.value.startsWith(version) && parent && index !== undefined)` it could work? but i'm getting further and further out of my depth here on if that would break anything else. I do like the `.startsWith(version)` part though

I am not in any sort of rush on this, but I also don't know if @ybiquitous wants me to make a PR to https://github.com/stylelint/changelog-to-github-release-action or not, but if so, I can do that too.

jondricek commented:
With all this effort... maybe I should actually use stylelint for the first time too. That's what I told myself I'd do this morning anyway. At least I installed the VS Code extension!

jondricek commented:
Ok. I'm not trying to scope creep, but I noticed that I broke 3 links in the Changelog with adding dates to the headers. Specifically the 8.0.0 release in 2017 linked to places in the Changelog itself, which no longer would work. So i fixed those 3 broken links to point to their respective GitHub Releases for what they're talking about instead of the Changelog. Instead of digging through all the many changes in the Changelog, here is that commit that I'm talking about here: https://github.com/stylelint/stylelint/pull/8260/commits/6f4b9c0365ea4b74fd5341e18af5e8fbeafed018

ybiquitous commented:
@jondricek Thank you for creating the PR. To verify the changes, I've created and run a tiny script:

```js
import { execSync } from 'node:child_process';
import process from 'node:process';
import { readFileSync } from 'node:fs';

const changelogPath = process.argv[2];
const changelogContent = readFileSync(changelogPath, 'utf8');
const versionsAndDates = changelogContent.matchAll(/^## (.+) - (\d{4}-\d{2}-\d{2})$/gm);

const releases = JSON.parse(
	execSync('gh release list --limit=1000 --json=tagName,publishedAt').toString(),
);

for (const [, version, date] of versionsAndDates) {
	for (const { tagName, publishedAt } of releases) {
		if (version === tagName) {
			const actualDate = publishedAt.split('T')[0];

			if (date !== actualDate) {
				console.error(
					`${version} has a different date ${date} from ${actualDate} (https://github.com/stylelint/stylelint/releases/tag/${version})`,
				);
			}
		}
	}
}
```

```sh-session
$ node check-changelog.mjs CHANGELOG.md
16.9.0 has a different date 2024-08-27 from 2024-08-28 (https://github.com/stylelint/stylelint/releases/tag/16.9.0)
16.3.0 has a different date 2024-03-23 from 2024-03-24 (https://github.com/stylelint/stylelint/releases/tag/16.3.0)
16.2.0 has a different date 2023-01-19 from 2024-01-19 (https://github.com/stylelint/stylelint/releases/tag/16.2.0)
15.10.1 has a different date 2023-07-05 from 2023-07-06 (https://github.com/stylelint/stylelint/releases/tag/15.10.1)
15.6.0 has a different date 2023-04-22 from 2023-04-23 (https://github.com/stylelint/stylelint/releases/tag/15.6.0)
15.5.0 has a different date 2023-04-15 from 2023-04-16 (https://github.com/stylelint/stylelint/releases/tag/15.5.0)
15.4.0 has a different date 2023-03-31 from 2023-04-01 (https://github.com/stylelint/stylelint/releases/tag/15.4.0)
14.10.0 has a different date 2022-08-10 from 2022-08-11 (https://github.com/stylelint/stylelint/releases/tag/14.10.0)
14.9.1 has a different date 2022-06-10 from 2022-06-11 (https://github.com/stylelint/stylelint/releases/tag/14.9.1)
14.8.2 has a different date 2022-05-03 from 2022-05-04 (https://github.com/stylelint/stylelint/releases/tag/14.8.2)
13.2.0 has a different date 2020-02-13 from 2020-02-14 (https://github.com/stylelint/stylelint/releases/tag/13.2.0)
9.0.0 has a different date 2018-02-08 from 2018-02-18 (https://github.com/stylelint/stylelint/releases/tag/9.0.0)
7.4.2 has a different date 2016-10-09 from 2016-10-10 (https://github.com/stylelint/stylelint/releases/tag/7.4.2)
6.0.3 has a different date 2016-04-20 from 2016-04-21 (https://github.com/stylelint/stylelint/releases/tag/6.0.3)
5.4.0 has a different date 2016-04-08 from 2016-04-09 (https://github.com/stylelint/stylelint/releases/tag/5.4.0)
5.3.0 has a different date 2016-04-06 from 2016-04-07 (https://github.com/stylelint/stylelint/releases/tag/5.3.0)
5.2.0 has a different date 2016-03-24 from 2016-03-25 (https://github.com/stylelint/stylelint/releases/tag/5.2.0)
5.1.0 has a different date 2016-03-17 from 2016-03-18 (https://github.com/stylelint/stylelint/releases/tag/5.1.0)
5.0.0 has a different date 2016-03-10 from 2016-03-11 (https://github.com/stylelint/stylelint/releases/tag/5.0.0)
4.3.4 has a different date 2016-02-16 from 2016-02-17 (https://github.com/stylelint/stylelint/releases/tag/4.3.4)
4.3.3 has a different date 2016-02-13 from 2016-02-14 (https://github.com/stylelint/stylelint/releases/tag/4.3.3)
4.2.0 has a different date 2016-01-31 from 2016-02-01 (https://github.com/stylelint/stylelint/releases/tag/4.2.0)
3.2.2 has a different date 2016-01-09 from 2016-01-10 (https://github.com/stylelint/stylelint/releases/tag/3.2.2)
3.2.0 has a different date 2015-12-31 from 2016-01-07 (https://github.com/stylelint/stylelint/releases/tag/3.2.0)
1.1.0 has a different date 2015-09-19 from 2015-09-20 (https://github.com/stylelint/stylelint/releases/tag/1.1.0)
```

I got the different dates from the actual release dates as above. Could you double-check them?

ybiquitous commented on .changeset/rewrite-changelog.mjs:
```diff
@@ -40,13 +40,18 @@ let latestVersion = undefined;
 let subHeader = false;
 
+// Get today's date in YYYY-MM-DD format
+const today = new Date();
+const formattedDate = today.toISOString().split('T')[0];
```
```suggestion
const today = new Date().toISOString().split('T')[0];
```

ybiquitous commented on .changeset/rewrite-changelog.mjs:
```diff
@@ -40,13 +40,18 @@ let latestVersion = undefined;
 		if (!latestVersion) {
 			latestVersion = line.replace('## ', '');
-			newLines.push(line);
+			// Add today's date to the latest version
+			newLines.push(`## ${latestVersion} - ${formattedDate}`);
```
```suggestion
			newLines.push(`## ${latestVersion} - ${today}`);
```

ybiquitous reviewed (commented):
[suggestion] Minor refactoring for simplicity. How about reducing a variable and comments a bit like this?

Mouvedia commented:
> @Mouvedia I see your comment, but I'm not exactly sure what to do on my branch about it. I'm happy to add in the section for
> 
> ```
> ## 16.0.0-0–16.0.0-3 - 2023-11-08 – 2023-11-27
> ...
> <content here>
> ...
> ```
> 
> for you, but it looks like that whole section is missing from the Changelog at the moment, right?
> 
> Also, other than manually doing that for historic entries, I'm not sure how to codify that into the script or anything.

Don't bother Ill eventually do it in a separate PR.

Mouvedia commented:
> ```js
> const releases = JSON.parse(
> 	execSync('gh release list --limit=1000 --json=tagName,publishedAt').toString(),
> );
> ```

@ybiquitous you should use `git for-each-ref --sort=creatordate --format "%(creatordate:short) %(refname:short)" refs/tags` instead.

ybiquitous commented:
@Mouvedia `creatordate`s are still different by the modified script:

```js
import { execSync } from 'node:child_process';
import process from 'node:process';
import { readFileSync } from 'node:fs';

const changelogPath = process.argv[2];
const changelogContent = readFileSync(changelogPath, 'utf8');
const versionsAndDates = changelogContent.matchAll(/^## (.+) - (\d{4}-\d{2}-\d{2})$/gm);

const releases = execSync(
	'git for-each-ref --sort=creatordate --format "%(refname:short) %(creatordate:short)" refs/tags',
)
	.toString()
	.split('\n')
	.filter(Boolean)
	.map((e) => e.split(' '));

for (const [, version, date] of versionsAndDates) {
	for (const [tagName, publishedAt] of releases) {
		if (version === tagName) {
			const actualDate = publishedAt.split('T')[0];

			if (date !== actualDate) {
				console.error(
					`${version} has a different date ${date} from ${actualDate} (https://github.com/stylelint/stylelint/releases/tag/${version})`,
				);
			}
		}
	}
}
```

```sh-session
$ node check-changelog.mjs CHANGELOG.md
16.10.0 has a different date 2024-10-11 from 2024-10-12 (https://github.com/stylelint/stylelint/releases/tag/16.10.0)
16.9.0 has a different date 2024-08-27 from 2024-08-28 (https://github.com/stylelint/stylelint/releases/tag/16.9.0)
16.3.0 has a different date 2024-03-23 from 2024-03-24 (https://github.com/stylelint/stylelint/releases/tag/16.3.0)
16.2.0 has a different date 2023-01-19 from 2024-01-19 (https://github.com/stylelint/stylelint/releases/tag/16.2.0)
15.10.1 has a different date 2023-07-05 from 2023-07-06 (https://github.com/stylelint/stylelint/releases/tag/15.10.1)
15.9.0 has a different date 2023-06-23 from 2023-06-22 (https://github.com/stylelint/stylelint/releases/tag/15.9.0)
15.7.0 has a different date 2023-06-05 from 2023-06-06 (https://github.com/stylelint/stylelint/releases/tag/15.7.0)
15.6.3 has a different date 2023-06-03 from 2023-06-02 (https://github.com/stylelint/stylelint/releases/tag/15.6.3)
15.6.0 has a different date 2023-04-22 from 2023-04-23 (https://github.com/stylelint/stylelint/releases/tag/15.6.0)
15.5.0 has a different date 2023-04-15 from 2023-04-16 (https://github.com/stylelint/stylelint/releases/tag/15.5.0)
14.14.1 has a different date 2022-11-03 from 2022-11-04 (https://github.com/stylelint/stylelint/releases/tag/14.14.1)
14.10.0 has a different date 2022-08-10 from 2022-08-11 (https://github.com/stylelint/stylelint/releases/tag/14.10.0)
14.9.1 has a different date 2022-06-10 from 2022-06-11 (https://github.com/stylelint/stylelint/releases/tag/14.9.1)
14.8.2 has a different date 2022-05-03 from 2022-05-04 (https://github.com/stylelint/stylelint/releases/tag/14.8.2)
9.0.0 has a different date 2018-02-08 from 2018-02-18 (https://github.com/stylelint/stylelint/releases/tag/9.0.0)
3.2.2 has a different date 2016-01-09 from 2016-01-10 (https://github.com/stylelint/stylelint/releases/tag/3.2.2)
```

E.g.

```sh-session
$ git grep 16.9.0 CHANGELOG.md
CHANGELOG.md:80: ## 16.9.0 - 2024-08-27

$ git for-each-ref --sort=creatordate --format "%(refname:short) %(creatordate:short)" refs/tags/16.9.0
16.9.0 2024-08-28
```

ybiquitous commented on .changeset/rewrite-changelog.mjs:
```diff
@@ -40,13 +40,18 @@ let latestVersion = undefined;
 			latestVersion = line.replace('## ', '');
-			newLines.push(line);
+			// Add today's date to the latest version
+			newLines.push(`## ${latestVersion} - ${formattedDate}`);
```
[suggestion] Considering a possible gap between a date running this script and an actual release date, it seems more proper to just leave a placeholder. For example:

```suggestion
			newLines.push(`## ${latestVersion} - YYYY-MM-DD ⚠️ Replace this placeholder with an actual release date and remove this warning`);
```

Any thoughts?

jeddy3 replied:
As we tend to merge a couple of final PRs before release, I think sometimes the date will be correct. Shall we assume it will be and add a note to [our release process](https://stylelint.io/maintainer-guide/releases/#:~:text=Release%20a%20Stylelint) as we already have an action there to check the order of the items within the changelog.

ybiquitous replied:
@jeddy3 Thanks for your feedback. It sounds better to me to add a step to check a release date before merging a "Prepare x.y.z" pull request. 👍🏼 

@jondricek Could you change the `releases.md` file? Probably around here 👇🏼 

https://github.com/stylelint/stylelint/blob/1709aa5773a53ec23d48c4be2f0238b315bd9481/docs/maintainer-guide/releases.md?plain=1#L12-L14

jondricek replied:
Done. (on my phone in the car right now, but looks right to me!)

ybiquitous replied:
Thank you, looks great 👍🏼

Mouvedia commented:
> @Mouvedia `creatordate`s are still different by the modified script:
> 
> ```js
> import { execSync } from 'node:child_process';
> import process from 'node:process';
> import { readFileSync } from 'node:fs';
> 
> const changelogPath = process.argv[2];
> const changelogContent = readFileSync(changelogPath, 'utf8');
> const versionsAndDates = changelogContent.matchAll(/^## (.+) - (\d{4}-\d{2}-\d{2})$/gm);
> 
> const releases = execSync(
> 	'git for-each-ref --sort=creatordate --format "%(refname:short) %(creatordate:short)" refs/tags',
> )
> 	.toString()
> 	.split('\n')
> 	.filter(Boolean)
> 	.map((e) => e.split(' '));
> 
> for (const [, version, date] of versionsAndDates) {
> 	for (const [tagName, publishedAt] of releases) {
> 		if (version === tagName) {
> 			const actualDate = publishedAt.split('T')[0];
> 
> 			if (date !== actualDate) {
> 				console.error(
> 					`${version} has a different date ${date} from ${actualDate} (https://github.com/stylelint/stylelint/releases/tag/${version})`,
> 				);
> 			}
> 		}
> 	}
> }
> ```
> 
> ```
> $ node check-changelog.mjs CHANGELOG.md
> 16.10.0 has a different date 2024-10-11 from 2024-10-12 (https://github.com/stylelint/stylelint/releases/tag/16.10.0)
> 16.9.0 has a different date 2024-08-27 from 2024-08-28 (https://github.com/stylelint/stylelint/releases/tag/16.9.0)
> 16.3.0 has a different date 2024-03-23 from 2024-03-24 (https://github.com/stylelint/stylelint/releases/tag/16.3.0)
> 16.2.0 has a different date 2023-01-19 from 2024-01-19 (https://github.com/stylelint/stylelint/releases/tag/16.2.0)
> 15.10.1 has a different date 2023-07-05 from 2023-07-06 (https://github.com/stylelint/stylelint/releases/tag/15.10.1)
> 15.9.0 has a different date 2023-06-23 from 2023-06-22 (https://github.com/stylelint/stylelint/releases/tag/15.9.0)
> 15.7.0 has a different date 2023-06-05 from 2023-06-06 (https://github.com/stylelint/stylelint/releases/tag/15.7.0)
> 15.6.3 has a different date 2023-06-03 from 2023-06-02 (https://github.com/stylelint/stylelint/releases/tag/15.6.3)
> 15.6.0 has a different date 2023-04-22 from 2023-04-23 (https://github.com/stylelint/stylelint/releases/tag/15.6.0)
> 15.5.0 has a different date 2023-04-15 from 2023-04-16 (https://github.com/stylelint/stylelint/releases/tag/15.5.0)
> 14.14.1 has a different date 2022-11-03 from 2022-11-04 (https://github.com/stylelint/stylelint/releases/tag/14.14.1)
> 14.10.0 has a different date 2022-08-10 from 2022-08-11 (https://github.com/stylelint/stylelint/releases/tag/14.10.0)
> 14.9.1 has a different date 2022-06-10 from 2022-06-11 (https://github.com/stylelint/stylelint/releases/tag/14.9.1)
> 14.8.2 has a different date 2022-05-03 from 2022-05-04 (https://github.com/stylelint/stylelint/releases/tag/14.8.2)
> 9.0.0 has a different date 2018-02-08 from 2018-02-18 (https://github.com/stylelint/stylelint/releases/tag/9.0.0)
> 3.2.2 has a different date 2016-01-09 from 2016-01-10 (https://github.com/stylelint/stylelint/releases/tag/3.2.2)
> ```
> 
> E.g.
> 
> ```
> $ git grep 16.9.0 CHANGELOG.md
> CHANGELOG.md:80: ## 16.9.0 - 2024-08-27
> 
> $ git for-each-ref --sort=creatordate --format "%(refname:short) %(creatordate:short)" refs/tags/16.9.0
> 16.9.0 2024-08-28
> ```

Thank you for the script.
Without it, it would have been hard to hunt the errors.

jondricek commented:
Ok, so to account for the date differentials, I have to admit that I manually entered them initially. But also, I grabbed the dates from NPM - not GitHub Releases. I updated the `check-changelog.mjs` below, and ran it with this output (after fixing some of my dates that were off). I will say that most of the dates that were off by one, which when stacked up against the npm release date was correct, but wrong vs. the GitHub Release. But this did catch 2 times where i was off by ten days and off by one year.

I also found that historically sometimes GitHub got released a day early, and sometimes npm got released a day early.

Anyway, what I've pushed to my latest branch should 100% match GitHub Release dates now.

This is blocked by me needing to add a test to the PR on the https://github.com/stylelint/changelog-to-github-release-action/pull/54. So I'll switch over there now

```
node check-changelog.mjs CHANGELOG.md
16.10.0 CHANGELOG: 2024-10-12 npm: 2024-10-11
15.9.0 CHANGELOG: 2023-06-22 npm: 2023-06-23
15.7.0 CHANGELOG: 2023-06-06 npm: 2023-06-05
15.6.3 CHANGELOG: 2023-06-02 npm: 2023-06-03
15.4.0 CHANGELOG: 2023-03-31 npm: 2023-04-01
14.14.1 CHANGELOG: 2022-11-04 npm: 2022-11-03
13.2.0 CHANGELOG: 2020-02-13 npm: 2020-02-14
7.4.2 CHANGELOG: 2016-10-09 npm: 2016-10-10
6.0.3 CHANGELOG: 2016-04-20 npm: 2016-04-21
5.4.0 CHANGELOG: 2016-04-08 npm: 2016-04-09
5.3.0 CHANGELOG: 2016-04-06 npm: 2016-04-07
5.2.0 CHANGELOG: 2016-03-24 npm: 2016-03-25
5.1.0 CHANGELOG: 2016-03-17 npm: 2016-03-18
5.0.0 CHANGELOG: 2016-03-10 npm: 2016-03-11
4.3.4 CHANGELOG: 2016-02-16 npm: 2016-02-17
4.3.3 CHANGELOG: 2016-02-13 npm: 2016-02-14
4.2.0 CHANGELOG: 2016-01-31 npm: 2016-02-01
Version 3.0.0-3.0.2 not found in npm registry.
2.3.7 CHANGELOG: 2015-12-09 npm: 2015-12-10
```

#### Contents of `check-changelog.mjs`

```
import { get } from 'node:https';
import { execSync } from 'node:child_process';
import process from 'node:process';
import { readFileSync } from 'node:fs';

const changelogPath = process.argv[2];
const changelogContent = readFileSync(changelogPath, 'utf8');
const versionsAndDates = changelogContent.matchAll(/^## (.+) - (\d{4}-\d{2}-\d{2})$/gm);

function getNpmReleaseDates(callback) {
    const url = 'https://registry.npmjs.org/stylelint';

    get(url, (res) => {
        let data = '';

        res.on('data', (chunk) => {
            data += chunk;
        });

        res.on('end', () => {
            const jsonData = JSON.parse(data);
            const releaseDates = jsonData.time;

            callback(releaseDates);
        });
    }).on('error', (err) => {
        console.error('Error fetching data:', err);
    });
}

const releases = execSync(
    'git for-each-ref --sort=creatordate --format "%(refname:short) %(creatordate:short)" refs/tags',
)
    .toString()
    .split('\n')
    .filter(Boolean)
    .map((e) => e.split(' '));

getNpmReleaseDates((npmReleaseDates) => {
    for (const [, version, changelogDate] of versionsAndDates) {
        let npmDateMismatch = false;
        let githubDateMismatch = false;
        let npmDate = '';
        let githubDate = '';

        if (npmReleaseDates[version]) {
            npmDate = npmReleaseDates[version].split('T')[0];
            if (changelogDate !== npmDate) {
                npmDateMismatch = true;
            }
        } else {
            console.error(`Version ${version} not found in npm registry.`);
        }

        for (const [tagName, publishedAt] of releases) {
            if (version === tagName) {
                githubDate = publishedAt.split('T')[0];
                if (changelogDate !== githubDate) {
                    githubDateMismatch = true;
                }
                break;
            }
        }

        if (npmDateMismatch || githubDateMismatch) {
            console.error(
                `${version} CHANGELOG: ${changelogDate}` +
                (npmDateMismatch ? ` npm: ${npmDate}` : '') +
                (githubDateMismatch ? ` GitHub: ${githubDate}` : '')
            );
        }
    }
});
```

ybiquitous commented:
@jondricek Thanks for the confirmation. Checking published dates to npm sounds more reasonable. 👍🏼 

Note: I happened to find that https://github.com/stylelint/stylelint/releases/tag/2.3.7 doesn't actually have a GitHub release. 😅 

So, would you mind fixing the following inconsistent dates between CHANGELOG and npm?

```
16.10.0 CHANGELOG: 2024-10-12 npm: 2024-10-11
15.9.0 CHANGELOG: 2023-06-22 npm: 2023-06-23
15.7.0 CHANGELOG: 2023-06-06 npm: 2023-06-05
15.6.3 CHANGELOG: 2023-06-02 npm: 2023-06-03
15.4.0 CHANGELOG: 2023-03-31 npm: 2023-04-01
14.14.1 CHANGELOG: 2022-11-04 npm: 2022-11-03
13.2.0 CHANGELOG: 2020-02-13 npm: 2020-02-14
7.4.2 CHANGELOG: 2016-10-09 npm: 2016-10-10
6.0.3 CHANGELOG: 2016-04-20 npm: 2016-04-21
5.4.0 CHANGELOG: 2016-04-08 npm: 2016-04-09
5.3.0 CHANGELOG: 2016-04-06 npm: 2016-04-07
5.2.0 CHANGELOG: 2016-03-24 npm: 2016-03-25
5.1.0 CHANGELOG: 2016-03-17 npm: 2016-03-18
5.0.0 CHANGELOG: 2016-03-10 npm: 2016-03-11
4.3.4 CHANGELOG: 2016-02-16 npm: 2016-02-17
4.3.3 CHANGELOG: 2016-02-13 npm: 2016-02-14
4.2.0 CHANGELOG: 2016-01-31 npm: 2016-02-01
Version 3.0.0-3.0.2 not found in npm registry.
2.3.7 CHANGELOG: 2015-12-09 npm: 2015-12-10
```

jondricek commented:
Haha, sure. I'll probably tackle it tomorrow though npm date tracking it is!

ybiquitous commented on docs/maintainer-guide/releases.md:
```diff
@@ -10,13 +10,14 @@
-   5. Select the version from the [`np`](https://github.com/sindresorhus/np) prompt that matches the one in the changelog.
-   6. Confirm the publishing of the package to [www.npmjs.com/package/stylelint](https://www.npmjs.com/package/stylelint).
-   7. Confirm the creation of the release at [stylelint/stylelint/releases](https://github.com/stylelint/stylelint/releases).
+   1. Ensure `CHANGELOG.md` has the correct date for the latest release.
```
[suggestion] One point. Can you include the UTC timezone? E.g.

```suggestion
   1. Ensure `CHANGELOG.md` has the correct date (UTC) for the latest release.
```

jondricek commented:
Alright - switched to npm published dates, and looks right to me!

With the PR over on https://github.com/stylelint/changelog-to-github-release-action/pull/54 finished, i think(?) this should be good to go. Unless that needs to be... released? To GitHub Actions Marketplace or some such first?

Anyway, Happy New Year, and let me know if there is anything else I can help with for this.

ybiquitous commented:
@jondricek Thank you for fixing the version!

The new `changelog-to-github-release-action` has been already available since we use its `main` branch:

https://github.com/stylelint/stylelint/blob/e3b3d36a1c3d00c3cc4577d5f934f828c346c344/.github/workflows/preparing-github-release.yml#L23

For the last thing, could you fix the lint error?

jondricek commented:
Looks like the lint error has been addressed successfully. Woo-hoo!

ybiquitous reviewed (approved):
Thanks, LGTM 👍

jeddy3 reviewed (approved):
@jondricek Thank you, LGTM.

It's great to see the dates. I think they'll help users see:

- our typical release cadence
- the longevity of the project

And thanks to everyone for checking dates and helping to refine the process.
